db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -2757,13 +2777,24 @@ Status DBImpl::WaitForFlushMemTables(
                  (flush_memtable_ids[i] != nullptr &&
                   cfds[i]->imm()->GetEarliestMemTableID() >
                       *flush_memtable_ids[i])) {
-        // Make file ingestion's flush wait until SuperVersion is also updated
-        // since after flush, it does range overlapping check and file level
-        // assignment with the current SuperVersion.
-        if (!flush_reason.has_value() ||
-            flush_reason.value() != FlushReason::kExternalFileIngestion ||
-            cfds[i]->GetSuperVersion()->imm->GetID() ==
-                cfds[i]->imm()->current()->GetID()) {
+        // Make file ingestion and checkpoint export flushes wait until
+        // SuperVersion is also updated. For file ingestion, after flush, it
+        // does range overlapping check and file level assignment with the
+        // current SuperVersion.
+        //
+        // For Checkpoint export flush, we wait until SuperVersion is updated to
+        // avoid a race condition where GetColumnFamilyMetaData observes the old
+        // version.
+        bool flush_reason_unknown = !flush_reason.has_value();
+        bool wait_for_superversion_update =
+            flush_reason.has_value() &&
+            (*flush_reason == FlushReason::kCheckpointExport ||
+             *flush_reason == FlushReason::kExternalFileIngestion);
+        bool superversion_updated = cfds[i]->GetSuperVersion()->imm->GetID() ==
```

Comment:
seems unnecessary to evaluate this in the case of the other flush reasons, i think doing it all in one if statement is probably cheaper even if its ungly

Reply:
Fair! It's such an ugly conditional though, couldn't resist :-)